File: backend/internal/model/password.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type PasswordStatus string

const (
	PasswordStatusPending  PasswordStatus = "pending"
	PasswordStatusActive   PasswordStatus = "active"
	PasswordStatusExpired  PasswordStatus = "expired"
	PasswordStatusDisabled PasswordStatus = "disabled"
)

type PasswordPolicy struct {
	ID           uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	PackageID    uuid.UUID      `gorm:"type:uuid;not null;index" json:"package_id"`
	PasswordHash string         `gorm:"type:varchar(255);not null" json:"-"`
	Priority     int            `gorm:"not null;default:1" json:"priority"`
	ValidFrom    *time.Time     `json:"valid_from,omitempty"`
	ValidUntil   *time.Time     `json:"valid_until,omitempty"`
	Status       PasswordStatus `gorm:"type:varchar(32);not null;default:'pending'" json:"status"`
	CreatedAt    time.Time      `gorm:"not null;default:now()" json:"created_at"`

	Package *FilePackage `gorm:"foreignKey:PackageID" json:"-"`
}

func (PasswordPolicy) TableName() string {
	return "password_policies"
}

type CreatePasswordRequest struct {
	Password   string     `json:"password" binding:"required,min=4"`
	Priority   int        `json:"priority" binding:"omitempty,min=1"`
	ValidFrom  *time.Time `json:"valid_from"`
	ValidUntil *time.Time `json:"valid_until"`
}

type BatchCreatePasswordRequest struct {
	Passwords []CreatePasswordRequest `json:"passwords" binding:"required,min=1,dive"`
}

type UpdatePasswordRequest struct {
	Password   *string    `json:"password"`
	Priority   *int       `json:"priority" binding:"omitempty,min=1"`
	ValidFrom  *time.Time `json:"valid_from"`
	ValidUntil *time.Time `json:"valid_until"`
}

type PasswordResponse struct {
	ID         uuid.UUID      `json:"id"`
	Password   string         `json:"password,omitempty"`
	Priority   int            `json:"priority"`
	ValidFrom  *time.Time     `json:"valid_from,omitempty"`
	ValidUntil *time.Time     `json:"valid_until,omitempty"`
	Status     PasswordStatus `json:"status"`
	CreatedAt  time.Time      `json:"created_at"`
}

func (p *PasswordPolicy) ToResponse(includePassword bool) PasswordResponse {
	resp := PasswordResponse{
		ID:         p.ID,
		Priority:   p.Priority,
		ValidFrom:  p.ValidFrom,
		ValidUntil: p.ValidUntil,
		Status:     p.Status,
		CreatedAt:  p.CreatedAt,
	}
	if includePassword {
		resp.Password = ""
	}
	return resp
}

type CurrentPasswordResponse struct {
	ID         uuid.UUID      `json:"id"`
	Password   string         `json:"password"`
	ValidFrom  *time.Time     `json:"valid_from,omitempty"`
	ValidUntil *time.Time     `json:"valid_until,omitempty"`
	Status     PasswordStatus `json:"status"`
}

func (p *PasswordPolicy) CalculateStatus() PasswordStatus {
	now := time.Now()

	if p.Status == PasswordStatusDisabled {
		return PasswordStatusDisabled
	}

	if p.ValidFrom != nil && now.Before(*p.ValidFrom) {
		return PasswordStatusPending
	}

	if p.ValidUntil != nil && now.After(*p.ValidUntil) {
		return PasswordStatusExpired
	}

	return PasswordStatusActive
}

File: backend/app/models/password.py
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Integer, Text, ForeignKey
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class PasswordStatus(str, enum.Enum):
    PENDING = "pending"
    ACTIVE = "active"
    EXPIRED = "expired"
    DISABLED = "disabled"


class PasswordPolicy(Base):
    __tablename__ = "password_policies"

    id = Column(String(36), primary_key=True)
    package_id = Column(String(36), ForeignKey("file_packages.id"), nullable=False, index=True)
    password_hash = Column(String(255), nullable=False)
    password_key_hash = Column(String(32), nullable=False)
    priority = Column(Integer, nullable=False, default=1)
    valid_from = Column(DateTime, nullable=True)
    valid_until = Column(DateTime, nullable=True)
    status = Column(String(32), nullable=False, default=PasswordStatus.PENDING.value)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    package = relationship("FilePackage", back_populates="passwords")

    def calculate_status(self) -> PasswordStatus:
        now = datetime.utcnow()
        
        if self.status == PasswordStatus.DISABLED.value:
            return PasswordStatus.DISABLED
        
        if self.valid_from and now < self.valid_from:
            return PasswordStatus.PENDING
        
        if self.valid_until and now > self.valid_until:
            return PasswordStatus.EXPIRED
        
        return PasswordStatus.ACTIVE
